### ISSUE TITLE
MySQL (yaSSL) returns unsupported cipher instead of Alert

### ISSUE THREAD

**Issue by sdann** [drwetter/testssl.sh#793]
yaSSL has a bug which will return a successful Server Hello message, with a cipher that the server supports, even if none of the client provided ciphers are supported.

This was first discussed in PR #782 and partially fixed by @dcooper16 in PR #784.

Opening a new Issue to track additional problems this server side bug causes.

I've also opened a bug with MySQL to fix the issue in their server at: https://bugs.mysql.com/87078

**Comment by sdann** [comment]
With #784 testssl.sh now detects if the server has returned an incorrect cipher and raises an error. Unfortunately, this short circuit behavior is causing some false successes to be reported. There may be several, but I've found the first in run_rc4().

With the default MySQL (yaSSL) configuration 2 RC4 ciphers are supported:

RC4-SHA
RC4-MD5

However, the -4 test currently returns "no RC4 ciphers detected (OK)" against this configuration.

This is because, before those 2 ciphers can be negotiated, the server bug is hit, with the client sending:

Cipher Suite: TLS_ECDHE_PSK_WITH_RC4_128_SHA (0xc033)

and the server responding with:

Cipher Suite: TLS_DHE_RSA_WITH_AES_128_CBC_SHA (0x0033)

Causing the loop to terminate with success.

Two possible fixes:
1) Iterate over every cipher 1 at a time.
    - much more expensive in # of connections
    - 1 off fix for this server issue
2) Detect tls_sockets() returns 1 and mark the test output as "UNKNOWN" or "SERVER ERROR".

**Comment by dcooper16** [comment]
Hi @sdann,

I think we'd need more information to understand what is going on here, as `run_rc4()` doesn't work as you describe. The client doesn't send TLS_ECDHE_PSK_WITH_RC4_128_SHA (0xc033), it sends a list of every RC4 cipher. So, if the client is sending a list of every RC4 cipher and the server responds with TLS_DHE_RSA_WITH_AES_128_CBC_SHA (0x0033), that either means:

- The server doesn't support any RC4 ciphers; or 

- The server is responding with a cipher not in the ClientHello even though it supports one of the ciphers in the ClientHello.

Also, note that `run_rc4()` first tests the server using OpenSSL's s_client and then tests it using `tls_sockets()`. The idea being that `tls_sockets()` is just used to find the ciphers that aren't supported by the version of OpenSSL being used. So, if you are using the OpenSSL that comes with testssl.sh or another OpenSSL that supports RC4-SHA and/or RC4-MD5, then the cipher(s) should have been detected by OpenSSL before testssl.sh even tries to find any RC4 ciphers using `tls_sockets()`.

Could you try running the following test against your server, and let us know what the output is?
```
bash testssl.sh --debug 5 --devel 03 "c0,11, c0,07, 00,66, c0,16, 00,18, c0,0c, c0,02, 00,05, 00,04, 00,92, 00,8a, 00,20, 00,24, c0,33, 00,8e, 00,65, 00,64, 00,60, 00,17, 00,03, 00,28, 00,2b, 00,ff" <URI>
```
This has testssl.sh perform a single test against the server using `tls_sockets()`. It sends a TLSv1.2 ClientHello that lists all of the RC4 ciphers. IIf that test doesn't result in an RC4 cipher being returned, try just the ciphers that you mentioned:
```
bash testssl.sh --debug 5 --devel 03 "00,05,00,ff" <URI>
```
According to https://www.wolfssl.com/wolfSSL/Products-yassl.html, yaSSL doesn't support TLSv1.2, so you could also try a TLSv1.1 ClientHello:
```
bash testssl.sh --debug 5 --devel 02 "00,05,00,ff" <URI>
```

**Comment by sdann** [comment]
> The server is responding with a cipher not in the ClientHello even though it supports one of the ciphers in the ClientHello.

This is what's happening. The server is adhering to its cipher order, which is:

DHE-RSA-AES256-SHA
DHE-RSA-AES128-SHA
AES128-RMD
DES-CBC3-RMD
DHE-RSA-AES256-RMD
DHE-RSA-AES128-RMD
DHE-RSA-DES-CBC3-RMD
AES256-SHA
RC4-SHA
RC4-MD5
DES-CBC3-SHA
DES-CBC-SHA
EDH-RSA-DES-CBC3-SHA
EDH-RSA-DES-CBC-SHA
AES128-SHA:AES256-RMD

Since DHE-RSA-AES128-SHA (0x0033), is second in its list, and is incorrectly matching to TLS_ECDHE_PSK_WITH_RC4_128_SHA (0xc033) which the client sent, then the client app never matches against the later ordered:

RC4-SHA
RC4-MD5

**Comment by sdann** [comment]
Here's the output of the first command, changed to TLSv1.1:

```
$ ./testssl.sh --debug=6 --openssl=/opt/openssl/openssl-chacha-dev/bin/openssl --ip=192.168.56.101 --nodns --starttls=mysql --devel 02 "c0,11, c0,07, 00,66, c0,16, 00,18, c0,0c, c0,02, 00,05, 00,04, 00,92, 00,8a, 00,20, 00,24, c0,33, 00,8e, 00,65, 00,64, 00,60, 00,17, 00,03, 00,28, 00,2b, 00,ff" sd-ubuntu-srv1.hollywood.local:3306

TLS_LOW_BYTE/HEX_CIPHER: 02/c0,11, c0,07, 00,66, c0,16, 00,18, c0,0c, c0,02, 00,05, 00,04, 00,92, 00,8a, 00,20, 00,24, c0,33, 00,8e, 00,65, 00,64, 00,60, 00,17, 00,03, 00,28, 00,2b, 00,ff
do_allciphers          = false
do_vulnerabilities     = false
do_beast               = false
do_lucky13             = false
do_breach              = false
do_ccs_injection       = false
do_ticketbleed         = false
do_cipher_per_proto    = false
do_crime               = false
do_freak               = false
do_logjam              = false
do_drown               = false
do_header              = false
do_heartbleed          = false
do_mx_all_ips          = false
do_pfs                 = false
do_protocols           = false
do_rc4                 = false
do_renego              = false
do_std_cipherlists     = false
do_server_defaults     = false
do_server_preference   = false
do_ssl_poodle          = false
do_tls_fallback_scsv   = false
do_sweet32             = false
do_client_simulation   = false
do_cipher_match        = false
do_tls_sockets         = true
do_mass_testing        = false
do_display_only        = false
URI:                   : sd-ubuntu-srv1.hollywood.local:3306

###########################################################
    testssl.sh       2.9dev from https://testssl.sh/dev/
    (6698de8 2017-07-14 10:39:08 -- )

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-chacha (1.0.2k-dev)" [~183 ciphers]
 on hollywood:/opt/openssl/openssl-chacha-dev/bin/openssl
 (built: "reproducible build, date unspecified", platform: "linux-x86_64")


sd-ubuntu-srv1.hollywood.local:3306
/
 Start 2017-07-17 10:22:36        -->> 192.168.56.101:3306 (sd-ubuntu-srv1.hollywood.local) <<--

 A record via            supplied IP "192.168.56.101"
 rDNS (192.168.56.101):  --
 Service set:            STARTTLS via MYSQL -- attention, this is experimental
sending client hello...
=== starting mysql STARTTLS dialog ===
=== just read banner ===
[
5.7.18-0ubuntu0.16.04.1�G2I:PS���P/?zLFNfQI'mysql_native_password
received server greeting
C:  ���!
initiated STARTTLS
=== finished mysql STARTTLS dialog with 0 ===
"\x16\x03\x01\x00\xb0\x01\x00\x00\xac\x03\x02\x54\x51\x1e\x7a\xde\xad\xbe\xef\x31\x33\x07\x00\x00\x00\x00\x00\xcf\xbd\x39\x04\xcc\x16\x0a\x85\x03\x90\x9f\x77\x04\x33\xd4\xde\x00\x00\x2e\xc0\x11\xc0\x07\x00\x66\xc0\x16\x00\x18\xc0\x0c\xc0\x02\x00\x05\x00\x04\x00\x92\x00\x8a\x00\x20\x00\x24\xc0\x33\x00\x8e\x00\x65\x00\x64\x00\x60\x00\x17\x00\x03\x00\x28\x00\x2b\x00\xff\x01\x00\x00\x55\x00\x0f\x00\x01\x01\x00\x23\x00\x00\x33\x74\x00\x00\x00\x0a\x00\x3e\x00\x3c\x00\x0e\x00\x0d\x00\x19\x00\x1c\x00\x1e\x00\x0b\x00\x0c\x00\x1b\x00\x18\x00\x09\x00\x0a\x00\x1a\x00\x16\x00\x17\x00\x1d\x00\x08\x00\x06\x00\x07\x00\x14\x00\x15\x00\x04\x00\x05\x00\x12\x00\x13\x00\x01\x00\x02\x00\x03\x00\x0f\x00\x10\x00\x11\x00\x0b\x00\x02\x01\x00"
reading server hello...
00000000  16 03 02 00 4a 02 00 00  46 03 02 02 cf ea 9d 57  |....J...F......W|
00000010  7f 51 74 7b 63 88 99 6f  b3 a1 ad 4c 0e 23 4e 0f  |.Qt{c..o...L.#N.|
00000020  3c 26 d1 89 df cb be 87  dd ee 23 20 8e 53 09 09  |<&........# .S..|
00000030  39 09 9d 17 4e 8e c4 07  ea 05 d9 94 a4 95 45 31  |9...N.........E1|
00000040  4f dd b3 6e f0 91 f0 60  63 36 91 78 00 33 00 16  |O..n...`c6.x.3..|
00000050  03 02 06 30 0b 00 06 2c  00 06 29 00 06 26 30 82  |...0...,..)..&0.|

160302004A02000046030202CFEA9D577F51747B6388996FB3A1AD4C0E234E0F3C26D189DFCBBE87DDEE23208E53090939099D174E8EC407EA05D994A49545314FDDB36EF091F0606336917800330016030206300B00062C000629000626308206223082040AA00302010202021000300D06092A864886F70D01010B05003081A2310B30090603550406130255533113301106035504080C0A57617368696E67746F6E310D300B060355040A0C04544E494831233021060355040B0C1A5448494820436572746966696361746520417574686F72697479311D301B06035504030C14544E494820496E7465726D656469617465204341312B302906092A864886F70D010901161C696E666F407468656E6574776F726B6973686F7374696C652E636F6D301E170D3137303632313232333133385A170D3138303730313232333133385A308188310B30090603550406130255533113301106035504080C0A57617368696E67746F6E3110300E06035504070C0753656174746C65310D300B060355040A0C04544E4948311A3018060355040B0C11544E4948205765622053657276696365733127302506035504030C1E73642D7562756E74752D737276312E686F6C6C79776F6F642E6C6F63616C30820122300D06092A864886F70D01010105000382010F003082010A0282010100E55ED76AB7027B5E711810A3DE840735C1BE9F03EB7F499A7AAB99C20B3DA8C2C3BB0E2B42E3856C408CB9384CD8127951410C47C482342B4FAB4D45F1E6AFA1025732E9DCD6536812FE9CE7A019EE8D836F6174AD291B7182C193D5ED45549F7742438D9C3F0BC4F27E710B65C5A67E1FB37035244963618F355D45AC686AAC99260AF59591E610DA07C84F147C2F4967C3CBA551EF9BF276BB035DFCFB6A4136FB7BFBF43251A4132B20604162F71FD9D509E8F7FD50335E1ADD3B8CB94E49B3B37B035B61011B580DAB69367FE3DE3A141F5E2A6A484B6F1CF50DC90DA4456A8F9BA4F1F137014580E06229CE2C1E7898068BD1A3ED52CC6F9A922654B7530203010001A38201783082017430090603551D1304023000301106096086480186F8420101040403020640303306096086480186F842010D042616244F70656E53534C2047656E65726174656420536572766572204365727469666963617465301D0603551D0E041604144F297F923897D1B88FE9F7251F8DDF7B072D40BD3081DA0603551D230481D23081CF80142D15F23746AD907C032A6D8167665F93A268B818A181B2A481AF3081AC310B30090603550406130255533113301106035504080C0A57617368696E67746F6E3110300E06035504070C0753656174746C65310D300B060355040A0C04544E494831233021060355040B0C1A544E494820436572746966696361746520417574686F726974793115301306035504030C0C544E494820526F6F74204341312B302906092A864886F70D010901161C696E666F407468656E6574776F726B6973686F7374696C652E636F6D82021000300E0603551D0F0101FF0404030205A030130603551D25040C300A06082B06010505070301300D06092A864886F70D01010B05000382020100808BFDE09615A95AE7B76038E24BEF66C9171BF1137136596E325399EE9AE45117142801F20C2417600EEB63E0F0565079EE4E1C04EA9EC39D67D0C4261D9747623C62925D5E20999C05CEAA8EB8420E69038F2C33955C507093D08F1E98E0AFFEA52F1374794FCF2E9941241E4F63F33DA61F6B408CA9366C93971FB4BC0096D4A2495DE4D59F675AC4937C99D4803E1344156B50FBCB8B62D9C6D20549713C242EBC078A6A8667EE260E64314E16343E03201A44D7668F9A53C54AF7F3B8D683D3053F8F179E46CBF6AB899D1A00B45CCDBCC14814C321C546ED7ECF417176F54C905424CE00CE0EF848990FE3DEAC5E458EB5BD7287C93242F741C1844F76B17E6A4E9B5E8F8997D48045F3380E267C726723AD242288BB342953B0B7F9595FB9EBFFE87BE86D44FB177B326FFF260BF3D3A655945A1999DA5F9A5F4A734928891D296D3EE4B2214DFCB50C3D26044482920DCC776C0C2CF0138C8F0D92DEF9C1C43209E6C313A105C7C5227711D1D29155A6B495CA83D9D67CA582B636229F4EBB34D38385DF8CFA42A15CF0E764978654EBE591E77953B38D7A990DCB99C80A38BE1CCBDABA4FBAB9485D6A318E9DF9069D6D452B136A5843CA7ED08AC1773598B278E7877E4AE7F37FC37C1BA0872BD88A65CDC9BC75F793EA0C5A209E6417F80BB50C4258B9EAF2EE7C288F34B958647ECB312ACA3C0551678BD54F45160302030D0C00030901008A5DFAC066764E61FACAC037575C6D3F830AA1F5F1E67F3CC6AFDA8B26E61A745E64CBE208F109E3AFBB54292D97F459E626831F55CD1B5755426CE7B7DA6ED86DEEB14FA4D7F541E1B40BE19816E2ED16CF187D3F25C38259BDF48F57CA3E19E4F544E0CC80B31091180D64590A43F7FCCA01E81404F2CDA92A3CF3A52A83D8669FC92CC94F44055E5E0047220AE6B087A5743BE4A3FC2DDC49F2E1800D06717A773AA966703BBA8D2E605A39F72DD3F553476E57130187F9DE4D2092BED71EE0200C60C8CA35587D3F59EEFB677D647D8E776C61448A8C4DF012D4A4EA177566496CCF1428C69A3C71FDB83A6CE3A3A6065AA6F07A0015A55A6466000585B7000105010028C4D28D360F2AEE32C6C8EEB5EC364BB36908D0CC4FBE54E877140DBA52B66ADDC3967E41B761E4160396A8F1FA467BBF7154F521452498B6778EBE855DD7359A975E20A466BE4B1E9D92F808230988F7F6A98B03BC4AEE6111069BB5DB1226E564DA30942998397D1FE8BEB7BE294F2CAEA229A4E270C81A7EBD97811D4912D4DFD67F126E28B532276203DBFB91244078B9C5A6C8539C4124199A9523782EAC026C425343EAEC7B9B29C4E0915B7139FA2EE423AB230A63A1515EAED438E11199652F2B8EB857096492A6EA6BF6E8C7EDE001CBB42A3767F161D1FEA6644D3803A234AFE8A2B997E652C41A9FEAC86E43A314563926A61EA0407792BB0B78010042E0176E905F42600439DCAC084E8893E5AB0058809A16E6860E0A38DF1D289EE364346E066A55DA495058A3C7CFE0F684F6C7D4BD046519EF37B53A36DE57BF926B465B4824A028CE8944E4EB4750F628C1488A13228F2B5E6D0109BAA0770168E684FADD1ABC9677A04387427936FF63D30D181226781B59C1ED24455EDE741E545EF23F8E55F092EE9B66B9B4080BDFF33281A8687A1DEB07F587908212CD33F7AE1C969042A3136D80E65E05C6DD701F15FA469460BDB0D6CFDD80D499C9FA184DE5CA4A5137FF61BBCC7BFE53191D3EFB3F5B7D52FC0BA9F30F1BF73D68D1CE7F85EEC5E144C373AF31D09A8B666E2C80B3FFE4D6EF16C7BAF19011F78416030200090D000005020102000016030200040E000000
TLS message fragments:
     protocol (rec. layer):  0x0302
     tls_content_type:       0x16 (handshake)
     msg_len:                74

     protocol (rec. layer):  0x0302
     tls_content_type:       0x16 (handshake)
     msg_len:                1584

     protocol (rec. layer):  0x0302
     tls_content_type:       0x16 (handshake)
     msg_len:                781

     protocol (rec. layer):  0x0302
     tls_content_type:       0x16 (handshake)
     msg_len:                9

     protocol (rec. layer):  0x0302
     tls_content_type:       0x16 (handshake)
     msg_len:                4

TLS handshake messages:
     handshake type:         0x02 (server_hello)
     msg_len:                70

     handshake type:         0x0B (certificate)
     msg_len:                1580

     handshake type:         0x0C (server_key_exchange)
     msg_len:                777

     handshake type:         0x0D (certificate_request)
     msg_len:                5

     handshake type:         0x0E (server_hello_done)
     msg_len:                0

TLS server hello message:
     tls_protocol:           0x0302
     tls_sid_len:            0x20 / = 32
     tls_hello_time:         0x02CFEA9D date: invalid date '47180445'
     tls_cipher_suite:       0x0033
     tls_compression_method: 0x00 (NONE)

The ServerHello specifies a cipher suite that wasn't included in the ClientHello.
  (returned 156 lines)
1

DEBUG (level 5): see files in /tmp/testssl.fjnDRw
```

**Comment by dcooper16** [comment]
Hi @sdann,

The bug in yaSSL is different than what I originally thought. We you said that
> yaSSL has a bug which will return a successful Server Hello message, with a cipher that the server supports, even if none of the client provided ciphers are supported.

I thought that yaSSL sent a ServerHello with a cipher not supported by the client if the ClientHello didn't include any ciphers supported by the server. If that had been the case, then #784 would have solved the problem.

Now I understand that the bug is very different (https://bugs.mysql.com/bug.php?id=87078):
> I've found a repeatable bug, where the yaSSL based MySQL server will compare only the last byte of each cipher sent in the Client Hello message.

That seems like a much more difficult bug to work around. I don't know how we can make testssl.sh provide "correct" results for yaSSL without either dramatically slowing down testing for all other servers or adding a lot of additional code just for this one server.

The best solution I can think of would be to try to add a test for this bug into the new function that I proposed adding in #791. Do you know if the server will return an alert rather than a ServerHello if ther is no match between the last byte of any of the cipher suites in the ClientHello and the last byte of any cipher suite supported by the server? For example, what is the response to:
```
./testssl.sh --debug=6 --openssl=/opt/openssl/openssl-chacha-dev/bin/openssl --ip=192.168.56.101 --nodns --starttls=mysql --devel 02 "00,d5, 00,ff" sd-ubuntu-srv1.hollywood.local:3306
```

**Comment by sdann** [comment]
Hey @dcooper16,

Yeah, my first understanding of the issue was a general cipher mismatch, but hitting it more and more the last byte pattern became clear. 

I agree, working around this bug, while still providing useful tests will be a lot of special case code and a lot more client-server calls. For now, I propose we just return some kind of SERVER ERROR/INDETERMINATE error for these tests, instead of SUCCESS.

I like your idea of doing the test once at the beginning (based off --starttles=mysql) then storing the outcome in a variable.

I think we can deterministically detect the server has this issue. MySQL (yaSSL) does not return an Alert message, but does reliably return a MySQL error when no cipher bytes match:

```
00000000  16 00 00 02 ff 13 04 23  30 38 53 30 31 42 61 64  |.......#08S01Bad|
00000010  20 68 61 6e 64 73 68 61  6b 65                    | handshake|
0000001a
```

**Comment by drwetter** [comment]
> ``#08S01Bad handshake``

oh well. That's certainly a good indicator.

I don't think either it is a good idea to add another workaround to deal with this reliably. Another check for implementation bugs is fine.

Still, I didn't get how clients succeeded to continue to talk with those cra^W err.. implementation.

**Comment by sdann** [comment]
@dcooper16 

>Also, note that run_rc4() first tests the server using OpenSSL's s_client and then tests it using tls_sockets(). The idea being that tls_sockets() is just used to find the ciphers that aren't supported by the version of OpenSSL being used. So, if you are using the OpenSSL that comes with testssl.sh or another OpenSSL that supports RC4-SHA and/or RC4-MD5, then the cipher(s) should have been detected by OpenSSL before testssl.sh even tries to find any RC4 ciphers using tls_sockets().

Looks like this was not happening due to #795. Only the small set (3) of SSLv2 RC4 cipher suites was being tested by openssl  in `run_rc4()` due to this bug.

**Comment by sdann** [comment]
To recap, the following tests (printed title) run into this bug on the server side:

- NULL ciphers (no encryption)
- Weak 128 Bit ciphers (SEED, IDEA, RC[2,4])
- Strong encryption (AEAD ciphers)
- PFS is offered
- Cipher order
- RC4 (CVE-2013-2566, CVE-2015-2808)
- Testing 359 ciphers

Here's the remediation I propose for each:

1) Fixed by adding "inconclusive" tag to output #796
- NULL ciphers (no encryption)
- Weak 128 Bit ciphers (SEED, IDEA, RC[2,4])
- Strong encryption (AEAD ciphers)

2) Fixed due to bug in other part of the code #795
- RC4 (CVE-2013-2566, CVE-2015-2808)

3) WONTFIX: while the full list of ciphers is not printed for these tests, at least 2 are, giving an idea that PFS is supported, and what the top preferred ciphers are. That's good enough for now.
- PFS is offered
- Cipher order
- Testing 359 ciphers

**Comment by sdann** [comment]
Good news. MySQL has fixed this upstream.

https://bugs.mysql.com/bug.php?id=87078

> Fixed in 5.5.58, 5.6.38, 5.7.20, 8.0.4.

**Comment by drwetter** [comment]
all merged.
